### PR TITLE
Internal refactoring and modernization

### PR DESCRIPTION
**Pull request by ricellis**
## Checklist

- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description

Merge `modernization` branch to `main`.

Fixes i320

## Approach

The work has previously been reviewed in: #643 #650 #652 #658 #659 #660 #663 #668 #669 #670 #671 #672 #673 #676 #677 #678 #679 #682 #691 #692 #695 #696 #698 #702 #706 #708 #711

The only additional changes in this PR are from rebasing on `main` and cleaning up `package-lock.json`

- [NEW] Included `time` in `restored` API events and CLI output.
- [NEW] Added metadata to backup files.
- [FIXED] Double output of errors to stderr when using CLI.
- [FIXED] Error for broken JSON in backup files.
- [FIXED] Wrong error code used for incomplete changes item.
- [FIXED] Error if the log file already exists when starting a new backup.
- [REMOVED] Dependency on `async` module.
- [REMOVED] Dependency on `tmp` module.
- [REMOVED] Unused request handling code.
- [IMPROVED] Increased tolerance to server and network errors when spooling changes (via cloudant-node-sdk changes follower).
- [IMPROVED] Avoided double parsing of JSON batches when resuming a backup.
- [IMPROVED] Resumed backups identification of incomplete backup lines during restore.
- [IMPROVED] Added line numbers to errors from reading backup or log files.
- [IMPROVED] Replace custom liner with Node built-in readline.
- [IMPROVED] Added warning that `--buffer-size` has no effect with `--resume`.
- [IMPROVED] Documentation about compatibility.
- [IMPROVED] Resolved linter warnings.
- [IMPROVED] Internal enhancements with promises and pipelines replacing callbacks and event emitters.
- [IMPROVED] Added some new test cases and improved test stability.
- [NOTE] Versions older than 2.10.0 cannot restore backups created with 2.10.0. See [compatibility note](README.md#compatibility-note-1).

## Schema & API Changes

- Included `time` in `restored` API events and CLI output.
- Added metadata to backup files.

## Security and Privacy

- "No change"

## Testing

- New tests added and tests refactored for modernization changes as described in PRs listed above.

## Monitoring and Logging

- Logging enhancements as outlined in PRs listed.

### CODE REVIEW

File: includes/request.js [hunk 1]
```diff
@@ -14,68 +14,44 @@
 'use strict';
```
[the remark this answers is not in the document]
Comment (ricellis) [reply includes/request.js:14]: done

File: test/app.js [hunk 2]
```diff
@@ -37,19 +37,24 @@ const validateArgs = async function(url, opts, errorValidationForAssertRejects)
   return assert.rejects(backupPromise(url, nullStream, opts), errorValidationForAssertRejects);
```
[the remark this answers is not in the document]
Comment (ricellis) [reply test/app.js:37]: Last commit here was Dec 19, so I think this is correct as 2023.

File: CHANGES.md [hunk 3]
```diff
@@ -1,3 +1,25 @@
...
+- [IMPROVED] Increased tolerance to server and network errors when spooling changes (via cloudant-node-sdk changes follower).
+- [IMPROVED] Avoided double parsing of JSON batches when resuming a backup.
+- [IMPROVED] Resumed backups identification of incomplete backup lines during restore.
+- [IMPROVED] Added line numbers to errors from reading backup or log files.
+- [IMPROVED] Replace custom liner with Node built-in readline.
+- [IMPROVED] Added warning that `--buffer-size` has no effect with `--resume`.
+- [IMPROVED] Documentation about compatibility.
+- [IMPROVED] Resolved linter warnings.
+- [IMPROVED] Internal enhancements with promises and pipelines replacing callbacks and event emitters.
+- [IMPROVED] Added some new test cases and improved test stability.
```
[the remark this answers is not in the document]
Comment (ricellis) [reply CHANGES.md]: done

File: includes/allDocsGenerator.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+// Copyright © 2023 IBM Corp. All rights reserved.
```
Comment (eiri) [inline includes/allDocsGenerator.js:1]: 2024?

File: includes/backupMappings.js [hunk 5]
```diff
@@ -0,0 +1,260 @@
+// Copyright © 2017, 2023 IBM Corp. All rights reserved.
```
Comment (eiri) [inline includes/backupMappings.js:1]: Here too. I'm not sure if we want year when actual file was created or when we merge it in `main`, but we do 2017, 2024 elsewhere

File: includes/parser.js [hunk 6]
```diff
@@ -124,6 +124,6 @@ function parseRestoreArgs() {
 }
 
 module.exports = {
-  parseBackupArgs: parseBackupArgs,
-  parseRestoreArgs: parseRestoreArgs
+  parseBackupArgs,
+  parseRestoreArgs
```
Comment (eiri) [inline includes/parser.js:128]: Missing copyright update?

**Review by eiri — APPROVED** [review]

**Review by emlaver — APPROVED** [review]

**Pull request merged**